### ISSUE TITLE
Reproducing with Best Parameters for Fine-Tuning?

### ISSUE THREAD

**Issue by orionw** [Nicolas-BZRD/EuroBERT#3]
Hi there and awesome work!

I was hoping to reproduce your scores for fine-tuning, do you happen to have the configs available? I am mostly looking at XNLI/GLUE tasks but others would be helpful also.

**Comment by hgissbkh** [comment]
Hi @orionw ,

Thanks for your message!
Below are the best hyperparameters we found for the **XNLI** and **PAWS-X** tasks.

**Base Training Hyperparameters:**

| Hyperparameter        | Value   |
|-----------------------|---------|
| Effective batch size  | 32      |
| Training steps        | 10,000  |
| Warmup ratio          | 0.1     |
| LR scheduler          | Linear  |
| Adam β₁               | 0.9     |
| Adam β₂               | 0.95    |
| Adam ε                | 1e-5    |
| Weight decay          | 0.1     |
| Seed                  | 42      |

We also incorporate early stopping at each epoch (with patience=1) based on validation accuracy.

**Best Learning Rates by Model & Dataset:**

| Model             | PAWS-X | XNLI |
|-------------------|-----------|----------|
| EuroBERT-210m    | 3.6e-5    | 3.6e-5   |
| EuroBERT-610m     | 4.6e-5    | 3.6e-5   |
| EuroBERT-2.1B    | 3.6e-5    | 2.8e-5   |

Hope this helps,
Cheers!

**Comment by hgissbkh** [comment]
Just a quick clarification, @orionw.
We set the number of training steps to 10,000 in our paper to ensure a fair comparison across all models, while keeping computational costs reasonable.
That being said, feel free to adjust this value if you're aiming to optimize performance further.
In any case, the hyperparameters I shared are exactly the ones we used in the paper.

**Comment by orionw** [comment]
Thanks all for the help, this is awesome! I'm encountering a bit of difficulty reproducing - by chance did you use the [run_xnli.py](https://github.com/huggingface/transformers/blob/main/examples/pytorch/text-classification/run_xnli.py) script from HF? And from the paper it seems like you are reporting validation metrics, correct? It could totally be on my end, so if this is correct I will go back to triple check / see if I missed something.

**Comment by hgissbkh** [comment]
We use our own evaluation pipeline, which is implemented in our GitHub repository. We're currently working on making it publicly available to improve reproducibility. If you don't mind waiting a little, it should be ready soon.

**Comment by Nicolas-BZRD** [comment]
Hey @orionw, can you share some details about your reproduction setup, issues? Do you have training logs and final metric performance?

**Comment by orionw** [comment]
Sorry for the delay, was doing a bit of traveling. Here are the commands I am using with a single GPU:

```
#!/bin/bash

MODEL_PATH=EuroBERT/EuroBERT-210m 
OUTPUT_DIR=results/eurobert-210m-reproduce
TASK=mnli

echo $OUTPUT_DIR
mkdir -p $OUTPUT_DIR

printf '\n\n\n\n TASK_NAME: '$TASK'\n'
python engine/run_glue.py \
    --model_name_or_path $MODEL_PATH \
    --trust_remote_code true \
    --ignore_mismatched_sizes false \
    --task_name $TASK \
    --do_train \
    --do_eval \
    --max_seq_length 128 \
    --per_device_train_batch_size 32 \
    --learning_rate 3.6e-5 \
    --warmup_ratio 0.1 \
    --weight_decay 0.1 \
    --max_steps 10000 \
    --adam_beta1 0.9 \
    --adam_beta2 0.95 \
    --adam_epsilon 1e-5 \
    --save_steps 10000 \
    --report_to none \
    --seed 42 \
    --output_dir $OUTPUT_DIR/$TASK


printf '\n\n\n\nTASK: xnli\n'
# do just validation
LANGUAGES="ar bg de el en es fr hi ru sw th tr ur vi zh"
for lang in $LANGUAGES; do
  echo "Evaluating language: $lang"
  python engine/run_xnli.py \
    --model_name_or_path $OUTPUT_DIR/mnli \
    --trust_remote_code true \
    --language "$lang" \
    --do_eval \
    --max_seq_length "128" \
    --per_device_eval_batch_size "128" \
    --report_to "none" \
    --output_dir "$OUTPUT_DIR/xnli/$lang"
done

# do predict
  python engine/run_xnli.py \
    --model_name_or_path $OUTPUT_DIR/mnli \
    --trust_remote_code true \
    --language "en" \
    --do_eval \
    --do_predict \
    --max_seq_length "128" \
    --per_device_eval_batch_size "128" \
    --report_to "none" \
    --output_dir "$OUTPUT_DIR/xnli/all"
```

I am using the HF scripts in the [mGTE repo for the comparison](https://github.com/izhx/nlu-evals/blob/main/engine/run_glue.py), which AFAIK are just the basic HF scripts.

I didn't add early stopping but the loss is going down the whole time:
```
[INFO|trainer.py:2405] 2025-03-17 15:49:07,129 >> ***** Running training *****
[INFO|trainer.py:2406] 2025-03-17 15:49:07,129 >>   Num examples = 392,702
[INFO|trainer.py:2407] 2025-03-17 15:49:07,129 >>   Num Epochs = 1
[INFO|trainer.py:2408] 2025-03-17 15:49:07,129 >>   Instantaneous batch size per device = 32
[INFO|trainer.py:2411] 2025-03-17 15:49:07,129 >>   Total train batch size (w. parallel, distributed & accumulation) = 32
[INFO|trainer.py:2412] 2025-03-17 15:49:07,129 >>   Gradient Accumulation steps = 1
[INFO|trainer.py:2413] 2025-03-17 15:49:07,129 >>   Total optimization steps = 10,000
[INFO|trainer.py:2414] 2025-03-17 15:49:07,129 >>   Number of trainable parameters = 212,358,915
{'loss': 0.8868, 'grad_norm': 39.411495208740234, 'learning_rate': 1.8e-05, 'epoch': 0.04}                                
{'loss': 0.7069, 'grad_norm': 24.342424392700195, 'learning_rate': 3.6e-05, 'epoch': 0.08}                                
{'loss': 0.6867, 'grad_norm': 60.417938232421875, 'learning_rate': 3.4e-05, 'epoch': 0.12}                                
{'loss': 0.6586, 'grad_norm': 19.699575424194336, 'learning_rate': 3.2e-05, 'epoch': 0.16}                                
{'loss': 0.6254, 'grad_norm': 64.60340118408203, 'learning_rate': 3e-05, 'epoch': 0.2}                                    
{'loss': 0.6018, 'grad_norm': 21.746938705444336, 'learning_rate': 2.8e-05, 'epoch': 0.24}                                
{'loss': 0.5832, 'grad_norm': 46.766448974609375, 'learning_rate': 2.6000000000000002e-05, 'epoch': 0.29}                 
{'loss': 0.554, 'grad_norm': 21.285343170166016, 'learning_rate': 2.4e-05, 'epoch': 0.33}                                 
{'loss': 0.5556, 'grad_norm': 14.108466148376465, 'learning_rate': 2.2000000000000003e-05, 'epoch': 0.37}                 
{'loss': 0.5293, 'grad_norm': 16.027828216552734, 'learning_rate': 2e-05, 'epoch': 0.41}                                  
{'loss': 0.519, 'grad_norm': 18.57399559020996, 'learning_rate': 1.8e-05, 'epoch': 0.45}                                  
{'loss': 0.5047, 'grad_norm': 8.018926620483398, 'learning_rate': 1.6e-05, 'epoch': 0.49}                                 
{'loss': 0.5026, 'grad_norm': 13.147716522216797, 'learning_rate': 1.4e-05, 'epoch': 0.53}                                
{'loss': 0.4771, 'grad_norm': 16.532758712768555, 'learning_rate': 1.2e-05, 'epoch': 0.57}                                
{'loss': 0.4658, 'grad_norm': 13.521305084228516, 'learning_rate': 1e-05, 'epoch': 0.61}                                  
{'loss': 0.4625, 'grad_norm': 13.611384391784668, 'learning_rate': 8e-06, 'epoch': 0.65}                                  
{'loss': 0.4458, 'grad_norm': 16.328594207763672, 'learning_rate': 6e-06, 'epoch': 0.69}                                  
{'loss': 0.4477, 'grad_norm': 12.562980651855469, 'learning_rate': 4e-06, 'epoch': 0.73}                                  
{'loss': 0.4295, 'grad_norm': 15.678597450256348, 'learning_rate': 2e-06, 'epoch': 0.77}                                  
{'loss': 0.4102, 'grad_norm': 22.849977493286133, 'learning_rate': 0.0, 'epoch': 0.81}                                    
```


I am getting these results:

MNLI
```
***** eval metrics *****
  epoch_mm                   =     0.8149
  eval_accuracy_mm           =     0.8407
  eval_loss_mm               =     0.4111
  eval_runtime_mm            = 0:00:17.24
  eval_samples_mm            =       9832
  eval_samples_per_second_mm =    570.138
  eval_steps_per_second_mm   =     71.267
```

XNLI validation
```
lang | score
-- | --
ar | 0.6385542169
bg | 0.5401606426
de | 0.7048192771
el | 0.4144578313
en | 0.8421686747
es | 0.7485943775
fr | 0.7329317269
hi | 0.5514056225
ru | 0.7004016064
sw | 0.4048192771
th | 0.4124497992
tr | 0.6361445783
ur | 0.3345381526
vi | 0.686746988
zh | 0.7040160643
AVERAGE | 0.603480589


```

XNLI test:
```
{
    "all_metrics": {
        "ar": {
            "language": "ar",
            "predict_accuracy": 0.6385229540918164,
            "predict_loss": 0.8029021620750427,
            "predict_model_preparation_time": 0.0022,
            "predict_runtime": 8.6293,
            "predict_samples": 5010,
            "predict_samples_per_second": 580.577,
            "predict_steps_per_second": 4.635
        },
        "bg": {
            "language": "bg",
            "predict_accuracy": 0.5445109780439121,
            "predict_loss": 0.9601048827171326,
            "predict_model_preparation_time": 0.0022,
            "predict_runtime": 10.5299,
            "predict_samples": 5010,
            "predict_samples_per_second": 475.79,
            "predict_steps_per_second": 3.799
        },
        "de": {
            "language": "de",
            "predict_accuracy": 0.6990019960079841,
            "predict_loss": 0.7018681168556213,
            "predict_model_preparation_time": 0.0022,
            "predict_runtime": 8.1842,
            "predict_samples": 5010,
            "predict_samples_per_second": 612.156,
            "predict_steps_per_second": 4.887
        },
        "el": {
            "language": "el",
            "predict_accuracy": 0.411377245508982,
            "predict_loss": 1.118746042251587,
            "predict_model_preparation_time": 0.0022,
            "predict_runtime": 10.634,
            "predict_samples": 5010,
            "predict_samples_per_second": 471.129,
            "predict_steps_per_second": 3.762
        },
        "en": {
            "language": "en",
            "predict_accuracy": 0.8373253493013972,
            "predict_loss": 0.422607421875,
            "predict_model_preparation_time": 0.0022,
            "predict_runtime": 5.6084,
            "predict_samples": 5010,
            "predict_samples_per_second": 893.306,
            "predict_steps_per_second": 7.132
        },
        "es": {
            "language": "es",
            "predict_accuracy": 0.7403193612774451,
            "predict_loss": 0.6314635276794434,
            "predict_model_preparation_time": 0.0022,
            "predict_runtime": 7.7416,
            "predict_samples": 5010,
            "predict_samples_per_second": 647.153,
            "predict_steps_per_second": 5.167
        },
        "fr": {
            "language": "fr",
            "predict_accuracy": 0.7325349301397206,
            "predict_loss": 0.6389726400375366,
            "predict_model_preparation_time": 0.0022,
            "predict_runtime": 8.382,
            "predict_samples": 5010,
            "predict_samples_per_second": 597.708,
            "predict_steps_per_second": 4.772
        },
        "hi": {
            "language": "hi",
            "predict_accuracy": 0.556686626746507,
            "predict_loss": 0.9283156394958496,
            "predict_model_preparation_time": 0.0022,
            "predict_runtime": 11.0239,
            "predict_samples": 5010,
            "predict_samples_per_second": 454.466,
            "predict_steps_per_second": 3.628
        },
        "ru": {
            "language": "ru",
            "predict_accuracy": 0.7017964071856287,
            "predict_loss": 0.6953274607658386,
            "predict_model_preparation_time": 0.0022,
            "predict_runtime": 8.8974,
            "predict_samples": 5010,
            "predict_samples_per_second": 563.088,
            "predict_steps_per_second": 4.496
        },
        "sw": {
            "language": "sw",
            "predict_accuracy": 0.4025948103792415,
            "predict_loss": 1.2066030502319336,
            "predict_model_preparation_time": 0.0022,
            "predict_runtime": 9.6834,
            "predict_samples": 5010,
            "predict_samples_per_second": 517.382,
            "predict_steps_per_second": 4.131
        },
        "th": {
            "language": "th",
            "predict_accuracy": 0.4167664670658683,
            "predict_loss": 1.0988340377807617,
            "predict_model_preparation_time": 0.0022,
            "predict_runtime": 10.7728,
            "predict_samples": 5010,
            "predict_samples_per_second": 465.059,
            "predict_steps_per_second": 3.713
        },
        "tr": {
            "language": "tr",
            "predict_accuracy": 0.6594810379241517,
            "predict_loss": 0.800744891166687,
            "predict_model_preparation_time": 0.0022,
            "predict_runtime": 7.4748,
            "predict_samples": 5010,
            "predict_samples_per_second": 670.254,
            "predict_steps_per_second": 5.351
        },
        "ur": {
            "language": "ur",
            "predict_accuracy": 0.3435129740518962,
            "predict_loss": 1.2028554677963257,
            "predict_model_preparation_time": 0.0022,
            "predict_runtime": 11.0508,
            "predict_samples": 5010,
            "predict_samples_per_second": 453.361,
            "predict_steps_per_second": 3.62
        },
        "vi": {
            "language": "vi",
            "predict_accuracy": 0.6950099800399202,
            "predict_loss": 0.7086825370788574,
            "predict_model_preparation_time": 0.0022,
            "predict_runtime": 7.4937,
            "predict_samples": 5010,
            "predict_samples_per_second": 668.563,
            "predict_steps_per_second": 5.338
        },
        "zh": {
            "language": "zh",
            "predict_accuracy": 0.7081836327345309,
            "predict_loss": 0.696287989616394,
            "predict_model_preparation_time": 0.0022,
            "predict_runtime": 6.7316,
            "predict_samples": 5010,
            "predict_samples_per_second": 744.249,
            "predict_steps_per_second": 5.942
        }
    },
    "predict_accuracy": 0.6058416500332668
}
```

Perhaps I am doing the wrong procedure - did you not fine-tune on MNLI like this?

**Comment by hgissbkh** [comment]
Hey @orionw, thanks for the detailed follow-up!
In our paper, we fine-tune on XNLI using a multilingual setup rather than a cross-lingual one, meaning that all languages are present in the training, validation, and test splits (https://huggingface.co/datasets/mteb/xnli).
Also, it might be worth increasing the maximum sequence length. You're currently using 128, but 256 or even 512 could be more appropriate, especially since our tokenizer tends to produce slightly longer sequences compared to others.
Lastly, I noticed you're evaluating on some languages that EuroBERT doesn't support (e.g., bg, el, sw, th, ur). In our experiments, we filter out any unsupported languages before evaluation.

**Comment by orionw** [comment]
Ah, thank you so much! I think those were the missing pieces and now all my questions are answered. Thank you all again for your time!
